frame/node-authorization/src/lib.rs

```diff
@@ -0,0 +1,862 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Node authorization pallet
+//!
+//! This pallet manages a configurable set of nodes for a permissioned network.
+//! Each node is dentified by a PeerId (i.e. Vec<u8>). It provides two ways to
+//! authorize a node,
+//!
+//! - a set of well known nodes across different organizations in which the
+//! connections are allowed.
+//! - users can claim the ownership for each node, then manage the connections of
+//! the node.
+//!
+//! A node must have an owner. The owner can additionally change the connections
+//! for the node. Only one user is allowed to claim a specific node. To eliminate
+//! false claim, the maintainer of the node should claim it before even starting the
+//! node. This pallet uses offchain worker to set reserved nodes, if the node is not
+//! an authority, make sure to enable offchain worker with the right CLI flag. The
+//! node can be lagged with the latest block, in this case you need to disable offchain
+//! worker and manually set reserved nodes when starting it.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::{
+	collections::btree_set::BTreeSet,
+	iter::FromIterator,
+	prelude::*,
+};
+use codec::Decode;
+use frame_support::{
+	decl_module, decl_storage, decl_event, decl_error,
+	debug, ensure,
+	weights::{DispatchClass, Weight},
+	traits::{Get, EnsureOrigin},
+};
+use frame_system::ensure_signed;
+
+type PeerId = Vec<u8>;
```

Comment:
Why not use the [new type pattern](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html) here for additional compile time safety?

Similar to what is done here:

https://github.com/paritytech/substrate/blob/f85fabd753c40041edcaac487f6f1ce259e8d11b/primitives/core/src/offchain/mod.rs#L193-L195

Might as well be worth depending on a single definition of `OpaquePeerId` across frame and offchain worker, but I am not familiar enough with this code base to say where that should live.

Hope I am not missing something.

Reply:
The `PeerId` here is not exactly same with `OpaquePeerId`, OpaquePeerId is scale encoded if I understand correctly, but here we are using plain bytes (non encoded).

Reply:
I am not sure I follow your reasoning @kaichaosun, could you expand? The scale encoding is only needed when crossing runtime or network boundaries, and in the end, encoding a `Vec<u8>` pretty much boils down to a `Vec<u8>`.

Reply:
In lasted commit, `OpaquePeerId` used [here](https://github.com/paritytech/substrate/pull/6996/files#diff-b21b2185f4606c4b1f60d5d583c92a64R187) in `OpaqueNetworkState` and [here](https://github.com/paritytech/substrate/pull/6996/files#diff-25a47086678523a88fb9514ddbfeabfbR40) in `node-authorization`, they contain different (but same type as `Vec<u8>`) data from user perspective, encoded and without encoding. I feels losing the necessary type for safety, not sure if it's a sensible concern.

I added some doc to clarify this though.

Reply:
I am not quite sure I follow. Are you implying that the `Vec<u8>` within an `OpaquePeerId(Vec<u8>)` is actually the Scale encoding of a `PeerId`?

As far as I can tell by the code proposed in this pull request this is resolved, thus marking it as such.